lib/active_model/serializer/fieldset.rb

```diff
@@ -18,14 +18,23 @@ def fields_for(type)
       attr_reader :raw_fields
 
       private
-
       def parsed_fields
+        # TODO: this is not very flexible.
+        # - we should probably support symbol, array, hash, etc
+        #
+        # JSONAPI::IncludeDirective supports these, but also changes the resulting structure.
+        # is it worth it to duplicate some of that functionality?
+        # or should IncludeDirective be modified to only allow one layer of options?
```

Comment:
I have a hack that brings fields functionality to the Attributes adapter (I use this for CSV generation)
```ruby
module ActiveModelSerializers
  module Adapter
    class Attributes < Base
      def serializable_hash(options = nil)
        options = serialization_options(options)
        options[:fields] ||= instance_options[:fields]
        hash = serializer.serializable_hash(instance_options, options, self)
        fields = JSONAPI::IncludeDirective.new(options[:fields]).to_hash
        apply_fields_whitelist(hash, fields)
      end

      # this is a little backwards, but it's needed until AMS has a more unified interface
      def apply_fields_whitelist(hash, fields)
        return hash.map{ |e| apply_fields_whitelist(e, fields) } if hash.is_a?(Array)

        result = {}
        fields_keys = fields.keys
        hash.each do |k, v|
          next unless fields_keys.include?(k)
          if v.is_a?(Hash)
            result[k] = apply_fields_whitelist(v, fields[k])
          else
            result[k] = v
          end
        end

        result
      end
    end
  end
end
```

this sort of thing def needs to be unified. 
As for unification timeline, I think I'll start this weekend, if not Monday, but I'd like to see #1857 merged, as well as this one, and the test refactor merged. :-)